c-dependencies/js-compute-runtime/builtins/request-response.cpp

```diff
@@ -2384,8 +2384,162 @@ bool Response::redirect(JSContext *cx, unsigned argc, JS::Value *vp) {
+  callbackCalled = true;
+  auto outstr = static_cast<std::u16string *>(out);
+  outstr->reserve(1 + outstr->length() + strlen);
+  outstr->append(str, strlen);
+
+  return true;
+}
+} // namespace
+
+bool Response::json(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "json", 1)) {
+    return false;
+  }
+  JS::RootedValue data(cx, args.get(0));
+  JS::RootedValue init_val(cx, args.get(1));
+  JS::RootedObject replacer(cx);
+  JS::RootedValue space(cx);
+
+  std::u16string out;
+  // 1. Let bytes the result of running serialize a JavaScript value to JSON bytes on data.
+  callbackCalled = false;
+  if (!JS::ToJSON(cx, data, replacer, space, &write_json_to_buf, &out)) {
+    return false;
+  }
+  if (!callbackCalled) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_RESPONSE_JSON_INVALID_VALUE);
+    return false;
+  }
+  out.shrink_to_fit();
+  // 2. Let body be the result of extracting bytes.
+
+  // 3. Let responseObject be the result of creating a Response object, given a new response,
+  // "response", and this’s relevant Realm.
+  JS::RootedValue status_val(cx);
+  uint16_t status = 200;
+
+  JS::RootedValue statusText_val(cx);
+  JS::RootedString statusText(cx, JS_GetEmptyString(cx));
+  JS::RootedValue headers_val(cx);
+
+  if (init_val.isObject()) {
+    JS::RootedObject init(cx, init_val.toObjectOrNull());
+    if (!JS_GetProperty(cx, init, "status", &status_val) ||
+        !JS_GetProperty(cx, init, "statusText", &statusText_val) ||
+        !JS_GetProperty(cx, init, "headers", &headers_val)) {
+      return false;
+    }
+
+    if (!status_val.isUndefined() && !JS::ToUint16(cx, status_val, &status)) {
+      return false;
+    }
+
+    if (status == 204 || status == 205 || status == 304) {
+      JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                                JSMSG_RESPONSE_NULL_BODY_STATUS_WITH_BODY);
+      return false;
+    }
+
+    if (!statusText_val.isUndefined() && !(statusText = JS::ToString(cx, statusText_val))) {
+      return false;
+    }
+
+  } else if (!init_val.isNullOrUndefined()) {
+    JS_ReportErrorLatin1(cx, "Response constructor: |init| parameter can't be converted to "
+                             "a dictionary");
+    return false;
+  }
+
+  fastly_response_handle_t response_handle = INVALID_HANDLE;
+  fastly_error_t err;
+  if (!fastly_http_resp_new(&response_handle, &err)) {
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+  if (response_handle == INVALID_HANDLE) {
+    return false;
+  }
+
+  auto make_res = HttpBody::make();
+  if (auto *err = make_res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  auto body = make_res.unwrap();
+  JS::RootedString string(cx, JS_NewUCStringCopyN(cx, out.c_str(), out.length()));
+  size_t encoded_len;
+  auto stringChars = encode(cx, string, &encoded_len);
```

Comment:
Not in this PR, but we should figure out a better api for `encode` to avoid the copy to `string`.